README.md
# lendingclub
API for lendingclub

## required:
pip install -r requirement.pip

## Install:
python setup.py install

## Usage 
- view Summary: 
```
from lendingclub import LendingClub
lc = LendingClub(investor_id=999999999, token='<auth_token>')
print lc.summary()
{u'accountTotal': 5107.35, u'receivedLateFees': 0, u'receivedInterest': 111.74, u'infundingBalance': 225,
u'outstandingPrincipal': 4831.93, u'investorId': 999999, u'receivedPrincipal': 518.07, u'accruedInterest': 24.79,
u'availableCash': 50.42, u'totalPortfolios': 3, u'totalNotes': 170}

```
- view current Note status:
```
notes = lc.get_detailednotes()
payments = lendingclub.NotesAnalytic(notes).estimate_daily_payments()
print '- income', '-' * 40
print json.dumps(payments, indent=3)
print '- stats', '-' * 40
result = lendingclub.NotesAnalytic(notes).report_note_stats()
print json.dumps(result, indent=3)
print '- sell list', '-' * 40
lendingclub.NotesAnalytic(notes).sell_list(x=10)

- income ----------------------------------------
{
   "2015-09-11T00:00:00.000-07:00": 7.056805555555554, 
   "2015-09-14T00:00:00.000-07:00": 0.85, 
   "2015-09-15T00:00:00.000-07:00": 9.400000000000002, 
   "2015-09-16T00:00:00.000-07:00": 5.262708333333334, 
   "2015-09-17T00:00:00.000-07:00": 10.260000000000002, 
   "2015-09-18T00:00:00.000-07:00": 1.66, 
   "2015-09-21T00:00:00.000-07:00": 19.51, 
   "2015-09-22T00:00:00.000-07:00": 6.31, 
   "2015-09-23T00:00:00.000-07:00": 4.9, 
   "2015-09-24T00:00:00.000-07:00": 15.669999999999998, 
   "2015-09-28T00:00:00.000-07:00": 3.21, 
   "2015-09-29T00:00:00.000-07:00": 12.780000000000001, 
   "2015-09-30T00:00:00.000-07:00": 22.819999999999997, 
   "2015-10-01T00:00:00.000-07:00": 17.21701388888889, 
   "2015-10-02T00:00:00.000-07:00": 2.17, 
   "2015-10-05T00:00:00.000-07:00": 1.56, 
   "2015-10-06T00:00:00.000-07:00": 3.1900000000000004, 
   "2015-10-07T00:00:00.000-07:00": 15.64, 
   "2015-10-08T00:00:00.000-07:00": 7.49, 
   "2015-10-21T00:00:00.000-07:00": 0
}
- stats ----------------------------------------
{
   "lastpaymentdate": {
      "34": [
         "loan_id=51947491&order_id=70156699&note_id=84487450"
      ], 
      "35": [
         "loan_id=44239461&order_id=61385146&note_id=76307547"
      ]
   }, 
   "paymentstatus": {
      "Scheduled": 10, 
      "Paid Off": 5, 
      "Completed": 156, 
      "Processing...": 3, 
      "null": 1, 
      "Overdue": 1
   }, 
   "loanstatus": {
      "Current": 159, 
      "Issued": 10, 
      "In Review": 1, 
      "Fully Paid": 5, 
      "In Grace Period": 1
   }
}
- maturity ----------------------------------------
{
   "00/36": 8, 
   "01/36": 24, 
   "02/36": 38, 
   "03/36": 3, 
   "04/36": 22, 
   "05/36": 65, 
   "09/36": 1, 
   "10/36": 1, 
   "16/36": 1, 
   "18/36": 1, 
   "00/60": 7, 
   "01/60": 2, 
   "03/60": 1, 
   "04/60": 1, 
   "17/60": 1
}

```
> auth_token is an authentication token generated via the Lending Club web application.

## Available Commands
```
from lendingclub import api
* api.summary()
* api.availablecash()
* api.notes()
* api.detailednotes()
* api.portfolios()
* api.loanlisting()
## Notes, Loans objects
* api.get_loans() 				# return list of loans
* api.get_notes()				# return list of notes
* api.get_detailednotes()		# return list of detailednotes
```

## Reference:
https://www.lendingclub.com/developers/lc-api.action
https://www.lendingclub.com/info/download-data.action

## Analysis
### estimate_daily_payment 
```
notes = lc.get_detailednotes()
payments = lendingclub.NotesAnalytic(notes).get_daily_payments()
print json.dumps(payments, indent=3)
{
   "2015-08-25T00:00:00.000-07:00": 4.65, 
   "2015-08-26T00:00:00.000-07:00": 5.57, 
   "2015-08-27T00:00:00.000-07:00": 15.866805555555555, 
   "2015-08-28T00:00:00.000-07:00": 9.082569444444445, 
   "2015-08-31T00:00:00.000-07:00": 5.9, 
   "2015-09-01T00:00:00.000-07:00": 8.17, 
   "2015-09-02T00:00:00.000-07:00": 1.56, 
   "2015-09-03T00:00:00.000-07:00": 3.95, 
   "2015-09-08T00:00:00.000-07:00": 15.23, 
   "2015-09-09T00:00:00.000-07:00": 0.84, 
   "2015-09-10T00:00:00.000-07:00": 1.53, 
   "2015-09-11T00:00:00.000-07:00": 8.58, 
   "2015-09-14T00:00:00.000-07:00": 1.68, 
   "2015-09-15T00:00:00.000-07:00": 9.400000000000002, 
   "2015-09-16T00:00:00.000-07:00": 0, 
   "2015-09-17T00:00:00.000-07:00": 9.430000000000001, 
   "2015-09-18T00:00:00.000-07:00": 1.66, 
   "2015-09-21T00:00:00.000-07:00": 41.449999999999996, 
   "2015-09-22T00:00:00.000-07:00": 6.31, 
   "2015-09-23T00:00:00.000-07:00": 0, 
   "2015-09-24T00:00:00.000-07:00": 5.6000000000000005
}
```

## Todo
- Data Analysis on old loans, find the pattern of defaults.
- generate static file to store summary, and render graph based on summary history
- filtered loan listing, with filter terms.
- Buy Notes (using loan ids)


CHANGES
1.0
* initialize project
* handle all get requests including
	Summary
	Available cash
	Notes owned
	Detailed notes owned
	Portfolio owned
	*Listed loans (all, or last avail list)

1.1
* edit summary
* browse secondary note
* sort/filter secondary note



lendingclub/api.py
#!/usr/bin/env python

import requests
import json
from objects import Note, DetailedNote, Loan, FolioLoan
requests.packages.urllib3.disable_warnings()


class RequstError(Exception):
    pass


class API:
    """ Lending Club API """
    def __init__(self, investor_id, token, version='v1'):
        self.api = 'https://api.lendingclub.com/api'
        self.investor_id = investor_id
        self.token = token
        self.version = version
        # summary cache
        self._summary = None
        # loans/notes cache for analytics
        self._notes = None
        self._loans = None
        self._detailednotes = None

    def _construct_base_url(self):
        return '{api}/investor/{version}/accounts/{investor_id}'.format(
            api=self.api,
            version=self.version,
            investor_id=self.investor_id)

    def _construct_loan_url(self):
        return '{api}/investor/{version}'.format(
            api=self.api,
            version=self.version)

    def get(self, endpoint, params={}, base=None):
        if not base:
            base = self._construct_base_url()
        url = '{base}/{endpoint}'.format(
            base=base,
            endpoint=endpoint)
        headers = {'Authorization': self.token,
                   'content-type': 'application/json'}
        response = requests.get(url,
                                headers=headers,
                                params=params)
        if response.status_code != 200:
            msg = '{}: {}'.format(response.status_code,
                                  response.content)
            raise RequstError(msg)
        return response.json()

    def summary(self):
        if not self._summary:
            self._summary = self.get('summary')
        return self._summary

    def availablecash(self):
        return self.get('availablecash')

    def notes(self):
        if not self._notes:
            self._notes = self.get('notes')['myNotes']
        return self._notes

    def detailednotes(self):
        if not self._detailednotes:
            self._detailednotes = self.get('detailednotes')['myNotes']
        return self._detailednotes

    def portfolios(self):
        return self.get('portfolios')

    def loanlisting(self, show_all=False):
        if show_all:
            params = {'showAll': True}
        else:
            params = {}
        loans = self.get(
            'loans/listing',
            params=params,
            base=self._construct_loan_url(),
        )
        self._loans = loans['loans']
        return loans

    def get_notes(self):
        """ Return Note Objects
        """
        if not self._notes:
            self.notes()
        return [Note(n) for n in self._notes]

    def get_detailednotes(self):
        """ Return List of DetailedNote Objects
        """
        if not self._detailednotes:
            self.detailednotes()
        return [DetailedNote(n) for n in self._detailednotes]

    def get_loans(self):
        """ Return List of Loan Objects
        """
        if not self._loans:
            self.loanlisting()
        return [Loan(l) for l in self._loans]


class FolioFn:
    """ FolioFnAPI 
        http://www.foliofnapi.com/api.html
        The FOLIOfn API project is a Moran Nachum project.

    """
    def __init__(self, user=None, password=None):
        self.api = 'http://www.foliofnapi.com/get.php'
        self._loans = None
        self.user = user
        self.password = password

    def _getbody(self, content):
        begin = '<body>'
        end = '</body>'
        idx1 = content.find(begin)
        idx2 = content.find(end)
        content = content[idx1+len(begin):idx2].strip()
        content = content.replace('": <br/>}', '": null}')
        content = content.replace('<br/>', '')
        content = content.replace("}, ]", "}]")
        return content

    def get(self, params={}):
        url = self.api
        params.update({'format': 'JSON'})
        headers = {'content-type': 'application/json'}
        response = requests.get(url,
                                headers=headers,
                                params=params)
        if response.status_code != 200:
            msg = '{}: {}'.format(response.status_code,
                                  response.content)
            raise RequstError(msg)
        content = self._getbody(response.content)
        if content == 'No rows returned.':
            return {'notes': []}
        return json.loads(content)

    def filter(self, params):
        if not self._loans:
            self._loans = self.get(params)['notes']
        return [FolioLoan(l) for l in self._loans]

    def sorted(self, loanlist):
        return sorted(loanlist, key=lambda k: k['loanClass'], reverse=True)

lendingclub/utils.py
from dateutil.parser import parse
import datetime
from pytz import timezone, UTC
import urllib
import urllib2
import cookielib

PROXY_FORMAT = '%Y-%m-%dT%H:%M:%S.%f%z'


def get_time_now():
    """
    This will return the current UTC time, which will include UTC info
    """
    return datetime.datetime.utcnow().replace(tzinfo=UTC)


def parse_time(date_str):
    if not date_str:
        return None
    return parse(date_str)


def auth(login_url, username, password):
    cj = cookielib.CookieJar()
    opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cj))
    params = {'login_url': '',
              'login_email': username,
              'login_password': password,
              'login_remember_me': 'on',
              'offeredNotListedPromotionFlag': ''}

    login_data = urllib.urlencode(params)
    resp = opener.open(login_url, login_data)
    print resp.read()
    return opener

lendingclub/analytic.py
#!/usr/bin/env python

from collections import OrderedDict
from utils import get_time_now, parse_time, auth


class Filter:

    @classmethod
    def lt(cls, ele, val):
        return ele < val

    @classmethod
    def gt(cls, ele, val):
        return ele > val

    @classmethod
    def lte(cls, ele, val):
        return ele <= val

    @classmethod
    def gte(cls, ele, val):
        return ele >= val

    @classmethod
    def eq(cls, ele, val):
        return ele == val

    @classmethod
    def contain(cls, ele, vals):
        return ele in vals

FILTER_MAPPING = {'<': Filter.lt,
                  '>': Filter.gt,
                  '>=': Filter.gte,
                  '<=': Filter.lte,
                  '==': Filter.eq,
                  }


class BaseAnalytic:

    def __init__(self):
        pass

    # def filter(self, seq, rule):
    #     """ common filter function
    #             rule = ('attr', '>', 30)
    #             attr, filter, value
    #     """
    #     a, f, v = seq
    #     fn = FILTER_MAPPING[f]
    #     return filter(lambda x: fn(x.__getattr__(a))


class NotesAnalytic(BaseAnalytic):

    def __init__(self, detailednotes):
        self.notes = detailednotes

    def estimate_daily_payments(self):
        """ return dict of dates, and expect payment
        """
        result = {}
        for note in self.notes:
            if note.loanStatus in ("Issued", "Current"):
                nxt_pmt_date = note.nextPaymentDate
                nxt_pmt_amt = note.estimate_next_payment_amount()
                if nxt_pmt_date:
                    result[nxt_pmt_date] = result.get(
                        nxt_pmt_date, 0) + nxt_pmt_amt
        ordered = OrderedDict(sorted(result.items(), key=lambda t: t[0]))
        return ordered

    def report_note_maturity(self):
        """ return
            {get_current_payment_number(): count}
        """
        result = {}
        for note in self.notes:
            cur_months = note.get_current_payment_number()
            cur_term = '{0:02d}/{1:02d}'.format(cur_months, note.loanLength)
            result[cur_term] = result.get(cur_term, 0) + 1

        ordered = OrderedDict(sorted(
            result.items(),
            key=lambda t: ''.join(t[0].split('/')[::-1])))

        return ordered

    def report_note_stats(self):
        """ return
            - loanstatus: count by loanstatus
            - paymentstatus: count by payment status
            - lastpayment date: date since last payment
                - 33 = 30 days + 3 processing date

        """
        loanstatus = {}
        paymentstatus = {}
        lastpaymentdate = {}
        for note in self.notes:
            loanstatus[note.loanStatus] = loanstatus.get(
                note.loanStatus, 0) + 1
            paymentstatus[note.currentPaymentStatus] = paymentstatus.get(
                note.currentPaymentStatus, 0) + 1
            if note.lastPaymentDate and note.loanStatus in ['Issued', 'Current']:
                day_since_last_pmt = note.days_since_last_pmt()
                if day_since_last_pmt >= 33: #
                    if day_since_last_pmt in lastpaymentdate:
                        lastpaymentdate[day_since_last_pmt].append(
                            note.url_params())
                    else:
                        lastpaymentdate[day_since_last_pmt] = [
                            note.url_params()]
        return {'loanstatus': loanstatus,
                'paymentstatus': paymentstatus,
                'lastpaymentdate': lastpaymentdate}

    def sell_list(self, x=10):
        """ Sell note if it reach (X) %
        """
        row_table = "{:<55} |" + "{:<10}" * 6
        for note in self.notes:
            if note.profitgain_sell():
                print row_table.format(note.url_params(),
                                       'profit',
                                       note.principalReceived,
                                       note.noteAmount,
                                       note.calculate_note_profit(),
                                       note.loanStatus,
                                       ''
                                       )
            if note.cutloss_sell():
                print row_table.format(note.url_params(),
                                       'cutloss',
                                       note.principalReceived,
                                       note.noteAmount,
                                       note.calculate_note_profit(),
                                       note.loanStatus,
                                       ' ({} days)'.format(
                    note.days_since_last_pmt())
                )


class LoansAnalytic(BaseAnalytic):

    def __init__(self, loans):
        self.loans = loans


lendingclub/objects.py
#!/usr/bin/env python

import json
import datetime
from utils import get_time_now, parse_time, auth
import requests


class AttrError(Exception):
    pass


class BaseObject(object):
    """ Object Base Class
    """

    def __init__(self, dict_obj):
        self.__dict__ = dict_obj

    def __getattr__(self, name):
        try:
            val = self.__dict__[name]
            if type(val) == float:
                return round(val, 2)
            else:
                return val
        except KeyError:
            raise AttrError('{} does not exists [{}]'.format(
                name,
                self.__dict__.keys()))

    def __getitem__(self, name):
        return self.__dict__.get(name, None)

    def dumps(self):
        print json.dumps(self.__dict__, indent=3)

    def is_pass_filter(self, crit={}):
        """ crit = [ {'grade__gt': 'D',
                      ''}
                    ]
        """


class Summary(BaseObject):
    pass


class Note(BaseObject):
    """ 
    Note Object (meta data of each note)
    {
            "loanId":11111,
            "noteId":22222,
            "orderId":33333,
            "interestRate":13.57,
            "loanLength":36,
            "loanStatus":"Late (31-120 days)",
            "grade":"C",
            "loanAmount":10800,
            "noteAmount":25,
            "paymentsReceived":5.88,
            "issueDate":"2009-11-12T06:34:02.000-08:00",
            "orderDate":"2009-11-05T09:33:50.000-08:00",
            "loanStatusDate":"2013-05-20T13:13:53.000-07:00"
    }
    """

    def __str__(self):
        return '<Note: {}>'.format(self.noteId)


class DetailedNote(BaseObject):

    """ Detail Note Object provide all note attributes
    {
       "grade": "C1", 
       "loanId": 55555555, 
       "interestRate": 12.29, 
       "accruedInterest": 0.21, 
       "creditTrend": "UP", 
       "portfolioName": "xxxxxxxx", 
       "loanStatus": "Current", 
       "loanAmount": 10000, 
       "canBeTraded": true, 
       "loanStatusDate": "2015-04-24T00:00:00.000-07:00", 
       "interestPending": 0, 
       "lastPaymentDate": "2015-07-30T11:53:21.000-07:00", 
       "principalReceived": 2.49, 
       "nextPaymentDate": "2015-09-01T00:00:00.000-07:00", 
       "orderId": 11111111111, 
       "portfolioId": 11111111111, 
       "currentPaymentStatus": "Completed", 
       "purpose": "Debt consolidation", 
       "paymentsReceived": 3.44, 
       "noteAmount": 25, 
       "noteId": 11111111111, 
       "principalPending": 22.51, 
       "issueDate": "2015-03-26T00:00:00.000-07:00", 
       "interestReceived": 0.95, 
       "loanLength": 36, 
       "orderDate": "2015-03-20T00:00:00.000-07:00"
    }
    """

    def __str__(self):
        return '<DetailedNote: {}>'.format(self.noteId)

    def get_current_payment_number(self):
        """ return number of current payment
            calculate by:
            = number of month of (now() - orderdate) / 30
        """
        issueDate = parse_time(self.issueDate)
        lastPaymentDate = parse_time(self.lastPaymentDate)
        if not lastPaymentDate:
            return 0
        payment_no = (lastPaymentDate - issueDate).days / 30
        if lastPaymentDate and not payment_no:
            return 1
        else:
            return payment_no

    def estimate_next_payment_amount(self):
        """ return estimate payment for this note
            calculate by:
            = paymentReceived / payment_no
        """
        payment_no = self.get_current_payment_number()

        if self.loanStatus == "Fully Paid":
            return 0
        elif self.loanStatus not in ("Issue", "Current"):
            return 0
        elif not self.lastPaymentDate:
            return self.noteAmount * (100 + self.interestRate) / 100 / self.loanLength
        return round(self.paymentsReceived / payment_no, 2)

    def calculate_note_profit(self):
        """ calculate note's profit in percent %
        """
        return round(float(self.paymentsReceived - self.principalReceived) / self.noteAmount * 100.0, 2)

    def profitgain_sell(self, x=10):
        """ We should sell note when borrower paid half of the principal
            or profit reach x=10 percent
        """
        return (
            self.loanStatus == 'Current' and
            self.principalReceived > self.noteAmount / 2 or
            self.calculate_note_profit() > x)

    def cutloss_sell(self):
        return (
            self.loanStatus not in ['Issued', 'Current', 'Fully Paid'])

    def days_since_last_pmt(self):
        if not self.lastPaymentDate:
            return None
        return (get_time_now() - parse_time(self.lastPaymentDate)).days

    def url_params(self):
        return 'loan_id={}&order_id={}&note_id={}'.format(self.loanId, self.orderId, self.noteId)


class Loan(BaseObject):

    """
    https://www.lendingclub.com/developers/listed-loans.action
    {
            "id":111111,
            "memberId":222222,
            "loanAmount":1750.0,
            "fundedAmount":25.0,
            "term":36,                      **
            "intRate":10.99,                **
            "expDefaultRate":3.5,           **
            "serviceFeeRate":0.85,          **
            "installment":57.29,
            "grade":"B",                    **
            "subGrade":"B3",                **
            "empLength":0,                  **
            "homeOwnership":"OWN",          **
            "annualInc":123432.0,           **
            "isIncV":"Requested",
            "acceptD":"2014-08-25T10:56:29.000-07:00",
            "expD":"2014-09-08T10:57:13.000-07:00",
            "listD":"2014-08-25T10:50:20.000-07:00",
            "creditPullD":"2014-08-25T10:56:18.000-07:00",
            "reviewStatusD":"2014-09-03T14:41:53.957-07:00",
            "reviewStatus":"NOT_APPROVED",  
            "desc":"Loan description",      **
            "purpose":"debt_consolidation",
            "addrZip":"904xx",              **
            "addrState":"CA",               **
            "investorCount":"",             **
            "ilsExpD":"2014-08-25T11:00:00.000-07:00",
            "initialListStatus":"F",
            "empTitle":"",
            "accNowDelinq":"",              **
            "accOpenPast24Mths":23,
            "bcOpenToBuy":30000,
            "percentBcGt75":23.0,           **
            "bcUtil":23.0,                  **
            "dti":0.0,                      **
            "delinq2Yrs":1,
            "delinqAmnt":0.0,
            "earliestCrLine":"1984-09-15T00:00:00.000-07:00",
            "ficoRangeLow":750,
            "ficoRangeHigh":754,
            "inqLast6Mths":0,
            "mthsSinceLastDelinq":90,
            "mthsSinceLastRecord":0,
            "mthsSinceRecentInq":14,
            "mthsSinceRecentRevolDelinq":23,
            "mthsSinceRecentBc":23,
            "mortAcc":23,
            "openAcc":3,
            "pubRec":0,
            "totalBalExMort":13944,
            "revolBal":1.0,
            "revolUtil":0.0,
            "totalBcLimit":23,
            "totalAcc":4,
            "totalIlHighCreditLimit":12,
            "numRevAccts":28,
            "mthsSinceRecentBcDlq":52,
            "pubRecBankruptcies":0,
            "numAcctsEver120Ppd":12,
            "chargeoffWithin12Mths":0,
            "collections12MthsExMed":0,
            "taxLiens":0,
            "mthsSinceLastMajorDerog":12,
            "numSats":8,
            "numTlOpPast12m":0,
            "moSinRcntTl":12,
            "totHiCredLim":12,
            "totCurBal":12,
            "avgCurBal":12,
            "numBcTl":12,
            "numActvBcTl":12,
            "numBcSats":7,
            "pctTlNvrDlq":12,
            "numTl90gDpd24m":12,
            "numTl30dpd":12,
            "numTl120dpd2m":12,
            "numIlTl":12,
            "moSinOldIlAcct":12,
            "numActvRevTl":12,
            "moSinOldRevTlOp":12,
            "moSinRcntRevTlOp":11,
            "totalRevHiLim":12,
            "numRevTlBalGt0":12,
            "numOpRevTl":12,
            "totCollAmt":12
    }
    """

    def __str__(self):
        return '<Loan: {}>'.format(self.id)


class FolioLoan(BaseObject):

    """
    {
        "loanId": "28102419",
        "noteId": "58215920",
        "orderId": "50738493",
        "accruedInterest": 31.24,
        "status": "Current",
        "askPrice": 1989.93,
        "markupDiscount": 7.70,
        "ytm": 20.73,
        "daysSinceLastPayment": 20,
        "creditScoreTrend": "UP",
        "ficoRange": "735-739",
        "dateTimeListed": "2015-08-21",
        "neverLate": "true",
        "loanClass": "G1",
        "loanMaturity": 60,
        "originalNoteAmount": 2000.00,
        "interestRate": 2.00,
        "principalInterest": 
    },
    """

    def __str__(self):
        return '<FolioFn: {}/{}/{}>'.format(self.loanId,
                                            self.orderId,
                                            self.noteId)

    @property
    def daysSinceLastPayment(self):
        return self.__dict__.get('daysSinceLastPayment', 0)

    def construct_url(self):
        """
        https://www.lendingclub.com/foliofn/browseNotesLoanPerf.action?showfoliofn=true&loan_id=1593342&order_id=3469444&note_id=14470297
        """
        base = "https://www.lendingclub.com/foliofn/browseNotesLoanPerf.action?showfoliofn=true&"
        return "{}loan_id={}&order_id={}&note_id={}".format(base,
                                                            self.loanId,
                                                            self.orderId,
                                                            self.noteId)

    def get_loanlisting(self, email, password):
        login_url = 'https://www.lendingclub.com/account/login.action'
        opener = auth(login_url, email, password)
        print opener.get(self.consolidation())
